### PR TITLE
Task/2015 part5 notifications with substitution

### PR DESCRIPTION
**Pull request by kzangeli**
Substitution part of the subscription templates for notifications.
Pretty much untested, but seems to work.
An important amount of functional tests to be added once the pipeline is complete and possible bug fixes, of course.
The temporary unit tests will also be removed once the pipeline is complete and functests possible.

### CODE REVIEW

File: test/unittests/mongoBackend/mongoNotifyContext_test.cpp [hunk 1]
```diff
@@ -980,3 +980,112 @@ TEST(mongoNotifyContextRequest, templateNotificationStdPayload)
 
   utExit();
 }
+
+
+
+/* ****************************************************************************
+*
+* templateNotificationOverrideStdHeader -
```
Comment (fgalan) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: To be removed after closing the end-to-end? Thus, it should be marked with "FIXME: to be removed" or something like that.

Comment (kzangeli) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: This was taken care of in the last PR.
Before the 3-4 temporal unittests we have the following comment:

    //
    // FIXME: All following tests (templateNotification*) must be removed before
    //        merging into develop.
    //
    //

File: src/lib/rest/httpRequestSend.cpp [hunk 2]
```diff
@@ -131,6 +131,42 @@ static char* curlVersionGet(char* buf, int bufLen)
   return buf;
 }
 
+
+
+/* ****************************************************************************
+*
+* httpHeaderAdd - 
+*/
```
Comment (fgalan) [inline src/lib/rest/httpRequestSend.cpp]: A little explanation about this functions parameter would help.

Comment (fgalan) [inline src/lib/rest/httpRequestSend.cpp]: ... I mean, as part of the function comment preamble.

Comment (kzangeli) [inline src/lib/rest/httpRequestSend.cpp]: Sure.
Actually, that is ALWAYS a good idea. 
Perhaps we should add it to the style guide ...

Comment (kzangeli) [inline src/lib/rest/httpRequestSend.cpp]: Fixed in 1c16405

File: test/unittests/mongoBackend/mongoNotifyContext_test.cpp (hunk 1 above)
Comment (fgalan) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: NTC

Comment (fgalan) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: NTC

File: src/lib/common/macroSubstitute.cpp [hunk 3]
```diff
@@ -31,9 +31,256 @@
...
+*
 * macroSubstitute - 
+*
+* An initial out-buffer of CHUNK_SIZE is allocated, and later on during the substitution if the
+* out-buffer is not big enough, it is realloced with another CHUNK_SIZE.
+*
+* 1024 is set as CHUNK_SIZE and hopefully for most substitutions it will be enough with ONE allocation.
+* Allocation is very costly operations ...
+* This function could be a lot faster is we took the first 1024 bytes from the stack instead of using calloc.
+* However, the function gets a little more complicated like that as the first reallos would have to be a normal malloc and a memcpy.
```
Comment (fgalan) [inline src/lib/common/macroSubstitute.cpp]: reallos?

File: src/lib/rest/httpRequestSend.cpp [hunk 4]
```diff
@@ -131,6 +131,53 @@ static char* curlVersionGet(char* buf, int bufLen)
...
 
+
+
+/* ****************************************************************************
+*
+* httpHeaderAdd - add HTTP header to the list of HTTP headers for curl
+*
+* PARAMETERS
+*   o headersP            pointer to the list of HTTP headers, to be used by curl
+*   o headerName          The name of the neader, e.g. "Content-Type"
```
Comment (fgalan) [inline src/lib/rest/httpRequestSend.cpp]: neader -> header

File: src/lib/rest/httpRequestSend.cpp [hunk 5]
```diff
@@ -131,6 +131,53 @@ static char* curlVersionGet(char* buf, int bufLen)
...
+*   o headerString        The complete header, e.g. "Content-Type: application/json"
+*   o headerTotalSizeP    Pointer to a variaqble holding the total size of the list of headers,
```
Comment (fgalan) [inline src/lib/rest/httpRequestSend.cpp]: variaqble -> variable

File: src/lib/common/macroSubstitute.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/common/macroSubstitute.cpp]: realloc :-)

Fixed in 9982847

File: src/lib/rest/httpRequestSend.cpp (hunk 5 above)
Comment (kzangeli) [inline src/lib/rest/httpRequestSend.cpp]: Fixed in 235f772

File: src/lib/rest/httpRequestSend.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/rest/httpRequestSend.cpp]: Fixed in 235f772

**Comment by fgalan** [comment]
LGTM (passing the ball to @crbrox for final LGTM + merge + delete branch)

**Comment by crbrox** [comment]
LGTM

**Pull request merged**